src/infra/core/src/repos/metadata_chain_impl.rs

```diff
@@ -720,14 +631,46 @@ where
         opts: AppendOpts<'a>,
     ) -> Result<Multihash, AppendError> {
         if opts.validation == AppendValidation::Full {
-            self.validate_append_prev_block_exists(&block).await?;
-            self.validate_append_sequence_numbers_integrity(&block)
-                .await?;
-            self.validate_append_seed_block_order(&block)?;
-            self.validate_append_system_time_is_monotonic(&block)
+            let hash = if let Some(b) = opts.precomputed_hash {
+                Cow::Borrowed(b)
+            } else {
+                let hash = self.meta_block_repo.get_block_hash(&block).int_err()?;
```

Comment:
In my previous comments I gave you two ideas:
- **calculate the hash of the block being appended** to be able to express all validation in `visit((hash, block): ...)` functions of visitors without making `hash` on `Option` which is annoying to use (especially for non-validating visitors where it would never be `None`)
- or make **append validating visitors accept the first block in constructors**

You have implemented BOTH pre-hashing and accepting new block in `::new()`, which I think is redundant. Constructors of validating visitors already know that this initial block is special and don't need to require a hash for it.

Reply:
Yes, I introduced the hash calculation because I mistakenly thought in one of the validators that we needed to know the hash of the added block

For this reason, it turned out that the two approaches were combined.

Hashing will be removed

Reply:
Done